include/envoy/stream_info/filter_state.h

```diff
@@ -55,6 +55,69 @@ class FilterState {
             (dynamic_cast<const T*>(getDataGeneric(data_name)) != nullptr));
   }
 
+  /**
+   * The addToList, hasListWithName, forEachListItem are variants of the above
+   * functions, that operate on list data. Multiple elements could be added
+   * the list under the same data_name.
+   * @param data_name the name of the data being added to the end of a list.
+   * @param data an owning pointer to the data to be stored.
+   * Note that data_names for list elements have no relation to the data_names
+   * for singleton data objects added through setData. All items added to the list
+   * must be of the same type.
+   */
+  template <typename T>
+  void addToList(absl::string_view data_name, std::unique_ptr<Object>&& data) {
+    const auto* list = getList(data_name);
+    if (list != nullptr) {
+      // Check type of first element in the list
+      const T* cast = dynamic_cast<const T*>(list->at(0).get());
+      if (!cast) {
+        throw EnvoyException(
+            fmt::format("List {} does not conform to the specified type", data_name));
+      }
+    }
+
+    addToListGeneric(data_name, std::move(data));
+  }
+
+  /**
+   * @param data_name the name of the list being probed.
+   * @return Whether a list of the type and name specified exists in the
+   * data store.
+   */
+  template <typename T> bool hasList(absl::string_view data_name) const {
+    const auto* list = getList(data_name);
+    return ((list != nullptr) && (dynamic_cast<const T*>(list->at(0).get()) != nullptr));
+  }
+
+  /**
+   * @param data_name the name of the list data being looked up.
+   * @param operation a lambda function that operates on each element in the list,
+   * if it exists. The iteration will stop if the lambda function returns false or
+   * reaches the end of the list. The iteration order will be the same as the order
+   * in which the elements were added to the list. Note that if an element in the
```

Comment:
nit, suggestion: "an element" -> "the elements in the list" (make it clearer they must have the same type).